src/sage/rings/burnside.py

```diff
@@ -0,0 +1,273 @@
+        self._names = dict() # dictionary mapping subgroups to names
+        Parent.__init__(self, category=Sets().Finite())
+
+    def _group_invariant(self, H):
+        return H.order()
+
+    def _normalize(self, H, name=None):
+        if not H.is_subgroup(self._G):
+            raise ValueError(f"{H} is not a subgroup of {self._G}")
+        p = self._group_invariant(H)
+        if p in self._cache:
+            for H0 in self._cache[p]:
+                if _is_conjugate(self._G, H, H0):
+                    if name is not None:
+                        self._names[H0] = name
+                    return H0
+            else:
+                self._cache[p].append(H)
+        else:
+            self._cache[p] = [H]
+        self._names[H] = name
+        return H
+
+    def _element_constructor_(self, x):
+        if x.is_subgroup(self._G):
+            return self.element_class(self, self._normalize(x))
+        else:
+            raise ValueError("x must be a subgroup of " + repr(self._G))
+
+        raise ValueError(f"unable to convert {x} into {self}")
+
+    def set_name(self, H, name):
+        r"""
+        Rename conjugacy class of ``H`` to ``name``.
+        Passing ``None`` to ``name`` will remove any previously assigned name.
+        """
+        if not isinstance(name, str):
+            raise TypeError("name must be a string")
+        H_norm = self._normalize(H)
+        self._names[H_norm] = name
+
+    def __iter__(self):
+        return iter(self(H) for H in self._G.conjugacy_classes_subgroups())
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+        """
+        return "Conjugacy classes of subgroups of " + repr(self._G)
+
+    Element = ConjugacyClassOfSubgroups
+
+class BurnsideRing(CombinatorialFreeModule):
+    def __init__(self, G, base_ring=ZZ):
+        """
+            sage: G = SymmetricGroup(4)
+            sage: B = BurnsideRing(G)
+            sage: TestSuite(B).run()
+        """
+        self._G = G
+        basis = ConjugacyClassesOfSubgroups(G)
+        basis.set_name(G, "1")
+        category = Algebras(base_ring).Commutative().WithBasis()
+        CombinatorialFreeModule.__init__(self, base_ring, basis, category=category)
+
+    def _element_constructor_(self, x):
+        r"""
+        Construct an element of the Burnside ring explicitly.
+
+        INPUT:
+
+        - ``x`` - data for an element
+
+        ``x`` can be a subgroup of `G` or a formal sum of such
+        subgroups.
+        """
+        if x in self.base_ring():
+            return x * self.one()
+
+        if isinstance(x, list) or isinstance(x, FormalSum):
+            # if x is a list of pairs (coeff, subgroup) or FormalSum
+            # Turn all subgroups into elements of ConjugacyClassesOfSubgroups
+            x = [(self._indices(subgroup), coeff) for coeff, subgroup in x]
+            return self._from_dict(dict(x))
+        elif x.is_subgroup(self._G):
```

Comment:
```suggestion
        if x.is_subgroup(self._G):
```
I think that this is not so useful, since we can simply construct an element using `sum`.